design/automatic-network-configuration.md

```diff
@@ -0,0 +1,355 @@
+      stored in `.status.hardware.nics`. CAPM3 will also create the
+      corresponding Interface CR if it is not created.
+- User:
+  - Creates production Network CRs.
+  - Creates a Cluster and specify its `.spec.networkRef`.
+    - CAPM3 finds a ready BMH.
+    - CAPM3 selects an Interface from the Interface CRs owned by this BMH to
+      connect to the network specified by the user.
+    - CAPM3 writes the chosen Interface CR information into the Network CR.
+    - Network Controller and Switch Controller will complete the automatic
+      configuration of the switch.
+    - After CAPM3 observes that the SwitchPort configuration is complete, it
+      triggers provisioning via the provisioning network.
+    - When the provisioning is completed, CAPM3 removes the bmh's Interface for
+      provisioning from the provisioning Network CR, thus disconnecting the BMH
+      from the provisioning network.
+
+### Changes to Current API
+
+#### Metal3Cluster CRD
+Add a `networkRef` field to `.spec` to point to the corresponding Network
+resource.
+
+```yaml
+Spec:
+ ...
+ networkRef: sample-network
+ ...
+```
+
+#### CAPM3
+- Before BareMetalHost provisioning, CAPM3 obtains all the NICs information. If
+  the corresponding Interface CR has not been created, it will create one.
+- If a BMH has multiple interfaces, CAPM3 selects one to use.
+- Writes the ref information of the selected Interface into `.spec.interfaces`
+  in the corresponding Network CR.
+- Checks the configuration status of the SwitchPort CR corresponding to the
+  selected Interface CR to ensure that the network configuration is successful,
+  and then proceed with the provisioning operation.
+
+#### BareMetal Operator
+- BMO obtains LLDP information (include ID of the port in switch and
+  information of vlan active on that port) which is collected by IPA during the
+  inspection process.
+- BMO stores the information of bare metal server’s NICs in the status of BMH
+  CR. The status of each NICs includes its name, mac address and LLDP
+  information.
+
+#### BMH CRD
+Add a field containing LLDP information in `.status.hardware.nics`.
+```yaml
+ status:
+ …
+ hardware:
+ …
+ nics:
+   // The following data is obtained by Baremetal Operator and
+   // filled with each lldp information.
+   - name: eth0
+     ...
+     lldp:
+   - name: eth1
+     ...
+```
+
+### Add new CRD and controller
+
+#### Interface CRD
+```yaml
+apiVersion: network.metal3.io/v1alpha1
+kind: Interface
+metaData:
+  name: bm0-eth0
+spec:
+  portID: 10
+  switchMacAddress: XX.XX.XX.XX
+  neighborRef:
+    - kind:
+      apiVersion:
+      namespace:
+      name:
+```
+
+#### Interface Controller
+- Watch
+  - Interface CR
+- Configuration Operations
+  - Whenever a new Interface CR is created, Interface Controller will find the
+    corresponding SwitchPort CR according to `.spec.portID` and
+    `.spec.switchMacaddress`, and write the reference of the SwitchPort CR into
+    `spec.neighborRef`.
```

Comment:
How would the controller know which switch this is the port of ? I think we should rather give the switch reference in the spec. We could get the switch name through lldp and match it with a switch CR, but we must consider cases where we do not have LLDP, so the linking must be done manually by the admin.